amy/emails/views.py

```diff
@@ -143,6 +155,15 @@ class ScheduledEmailUpdate(OnlyForAdminsMixin, FlaggedViewMixin, AMYUpdateView):
     model = ScheduledEmail
     object: ScheduledEmail
 
+    # Will lock this object in the database for the duration of the request.
+    # Most specifically, we want to lock it when we're saving the form. This way the DB
+    # helps us make sure the data is consistent.
+    # Additionally, we're limiting the queryset to only those objects that can be edited
+    # (see ScheduledEmailStatusActions).
+    queryset = ScheduledEmail.objects.filter(
+        state__in=ScheduledEmailStatusActions["edit"]
+    ).select_for_update()
```

Comment:
I wondered if this would lock the object as soon as the user navigates to the Update page, but this appears not to be the case (I assume I misunderstood how `select_for_update()` works):
In one tab I opened the Update view for a scheduled email, then in another tab I cancelled that email. When I submitted the update, I got a 404 error because the email can no longer be edited.

Could we add a check during the form validation stage for these views that shows an error to the user if the object no longer has the right status for the operation? (can be a separate issue)

Reply:
`select_for_update()` is a Django way of creating a `SELECT ... FOR UPDATE` PostgreSQL query. It's one of the methods of per-row explicit locking: https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS

This locks the row for the duration of a single transaction, and I wanted to use this because there's a risk someone will be updating an entry that simultaneously was fetched by the email worker and made "locked" (as in the scheduled email status).

You will have hard time reproducing this one on your local machine, the best way is to have a command that spawns multiple clients trying to do one single edit. I've had some very bad experience with this kind of behavior which resulted in a race condition that was difficult to catch.


> Could we add a check during the form validation stage for these views that shows an error to the user if the object no longer has the right status for the operation? (can be a separate issue)

That is possible, but I don't know how (yet). I'll add a low-priority issue for this.

Reply:
Ticket created: https://github.com/carpentries/amy/issues/2531